### ISSUE TITLE
?

### ISSUE THREAD

**Issue by congmonkey** [bob48523/modified_wrn#1]
class PreActBottleneck_p(nn.Module):
    '''Pre-activation version of the original Bottleneck module.'''
    widen_factor = 4
    
    def __init__(self, in_planes, planes, cardinality, stride=1):
        super(PreActBottleneck_p, self).__init__()
        
        self.inplanes= in_planes
        self.bn1 = nn.BatchNorm2d(in_planes)
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=1, bias=False)
        
        self.bn2 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride, padding=1, groups=cardinality, bias=False)
        
        self.bn3 = nn.BatchNorm2d(planes)
        self.conv3 = nn.Conv2d(planes, planes*self.widen_factor, kernel_size=1, bias=False)   
        
        self.bn4 = nn.BatchNorm2d(planes*self.widen_factor)

        if stride != 1:
            self.shortcut = nn.Sequential(
                nn.AvgPool2d(kernel_size=2, stride=2)             
            )
            
            

    def forward(self, x):
        global FLOPS
        out = self.bn1(x)
        shortcut = self.shortcut(x) if hasattr(self, 'shortcut') else x
        
        out = self.conv1(out)  
        FLOPS += self.conv1.in_channels*self.conv1.out_channels*self.conv1.kernel_size[0]*self.conv1.kernel_size[1]*out.size(2)*out.size(3)
        
        out = self.conv2(F.relu(self.bn2(out)))
        temp = self.conv2.in_channels*self.conv2.out_channels*self.conv2.kernel_size[0]*self.conv2.kernel_size[1]*out.size(2)*out.size(3)
        FLOPS += temp // self.conv2.groups        

        out = self.conv3(F.relu(self.bn3(out)))
        FLOPS += self.conv3.in_channels*self.conv3.out_channels*self.conv3.kernel_size[0]*self.conv3.kernel_size[1]*out.size(2)*out.size(3) 
        
        out = self.bn4(out)
        out = torch.cat([shortcut[:,:self.inplanes,:,:]+out[:,:self.inplanes,:,:], out[:,self.inplanes:,:,:]], 1)
        return out
